### ISSUE TITLE
Changing requirejs baseUrl has side-effects

### ISSUE THREAD

**Issue by Thanood** [aurelia/cli#387]
**I'm submitting a bug report**
(but I'm not entirely sure if this is a bug or intended)
- **Library Version:**
  0.21.0

**Please tell us about your environment:**
- **Operating System:**
  Windows 8
- **Node Version:**
  6.2.2
  
  
- **NPM Version:**
  3.10.5
  
  
- **Browser:**
  all
- **Language:**
  all

**Current behavior:**

According to [the documentation](http://aurelia.io/hub.html#/doc/article/aurelia/framework/latest/the-aurelia-cli/7) we can set `requirejs` config settings. I've tried to set this:

```
"baseUrl": "scripts/"
```

The reason for this is I'd like to dynamically load some components (aka "plugins") which are not bundled. I'd like to have these components in `scripts/plugins/`. If I don't change the baseUrl, the bundle tries to load these plugins from `src/plugins`, which will not be available on a production server.

If I do change the baseUrl, the module ids are generated to be relative to `scripts`, so for instance `main.js` is bundled like this:

``` javascript
define('../src/main', ...
```

This leads to the bootstrapper not finding any aurelia libs or the main entry point of the app.
Other files are bundled with the same prefix (`../src/`).

Without this change, requirejs modules are defined like this:

``` javascript
define('main', ...
```

**Expected/desired behavior:**

- **What is the expected behavior?**

In my opinion, setting `requirejs` baseUrl should _only_ affect the `requirejs` configuration in `vendor-bundle.js`.

In other words:
If I change this manually in `vendor-bundle.js` it works as I expect it. Files are bundled correctly and plugins are loaded from `scripts/plugins`.

I will upload a reproduction repo in a few minutes.

**Comment by Thanood** [comment]
### Reproduction repo

#### Without modifications, plugins loaded from `src/plugins`

* Clone this repo: https://github.com/Thanood/au-baseurl-issue
* ' au run`
* open the url in the browser and open devTools
* click "load plugins"
* receive a 404 error when loading `http://localhost:9000/src/plugins/index.js `

#### manually change `baseUrl`

* set `baseUrl` in `vendor-bundle.js` to `scripts/`
* reload the page
* receive a console log "plugin index loading" as well as a return value from the requirejs module (`Object {loaded: true}`)

#### change `baseUrl` in `aurelia.json`

* add this to build.loader in aurelia.json:

  ```
        "config": {
        "baseUrl": "scripts/"
      },
  ```

* `au run`
* see the app fail loading. :smiley:

**Comment by SaucyJack** [comment]
This issue is causing problems for us as well. Has anyone come up with a temporary work around? I am having to edit my vendor-bundle every time I build/run and manually change the `requirejs.config({"baseUrl"})`. Would it at least be possible to add a CLI build task to run a regex replace on this setting, and if so, how would that look?

**Comment by JeroenVinke** [comment]
@SaucyJack could you explain what you're trying to achieve? Perhaps a path mapping will suffice such as https://github.com/JeroenVinke/cli-baseurl/blob/master/aurelia_project/aurelia.json#L45

**Comment by Alexander-Taran** [comment]
workaround offered.
no activity since June 2017

**Comment by Alexander-Taran** [comment]
@3cp can be closed?

**Comment by 3cp** [comment]
It looks like implemented after this issue, but now it's broken again.

The baseUrl setting in the doc doesn't work. I will see if I can find out the regression.
https://aurelia.io/docs/cli/cli-bundler/basics#setting-the-baseurl

BTW, with dumber bundler (the accessor of cli-bundler), there are simple setting for `"src"` folder and `"baseUrl"` path. Which does exactly what you want with much cleaner generated requirejs config.

https://dumber.js.org

You will see in skeleton:
```
const dr = dumber({
  // src folder is by default "src".
  // src: 'src',

  // requirejs baseUrl, dumber default is "/dist"
  baseUrl: '/' + outputDir,
```

**Comment by 3cp** [comment]
The cli implementation only uses `project.paths.root` (which is set to "src") as baseUrl. All other paths in requirejs is relative to this root.

I am not going to "fix" this issue, as it will introduce a breaking change. Please consider using dumber bundler if need more flexibility. (Our Aurelia 2 skeleton supports dumber too, we deprecated cli-bundler for Aurelia 2).

```
this.loaderConfig = {
      baseUrl: project.paths.root,
      paths: ensurePathsRelativelyFromRoot(project.paths || {}),
      packages: [],
      stubModules: [],
      shim: {}
    };
```

In aurelia.json
```
"paths": {
    "root": "src",
    "resources": "resources",
    "elements": "resources/elements",
    "attributes": "resources/attributes",
    "valueConverters": "resources/value-converters",
    "bindingBehaviors": "resources/binding-behaviors"
  },
```

**Comment by 3cp** [comment]
BTW, I recall when I uses cli-bundler yrs before, I did use gulp-replace to touch the `vendor-bundle.js`, just replace `"baseUrl": "src"` to `"baseUrl": REQUIREJS_BASE_URL`.

I then have the global var set in html file like this
`<script>var REQUIREJS_BASE_URL="/scripts";</script>` which is easy for me to control.

BTW, dumber supported that global var by default.
Here is the generated skeleton from dumber:
```
requirejs.config({
  "baseUrl": (typeof REQUIREJS_BASE_URL === "string") ? REQUIREJS_BASE_URL : "/dist",
```

**Comment by Alexander-Taran** [comment]
@3cp awesome!!
